src/main/java/picard/illumina/parser/readers/CbclReader.java

```diff
@@ -0,0 +1,315 @@
+                int read = stream.read(byteBuffer.array());
+
+                //we need to read the first 6 bytes to determine the header size
+                if (read != INITIAL_HEADER_SIZE) {
+                    close();
+                    throw new RuntimeIOException(String.format("BCL %s has invalid header structure.", bclFile.getAbsoluteFile()));
+                }
+
+                short version = byteBuffer.getShort();
+                int headerSize = byteBuffer.getInt();
+
+                final ByteBuffer headerBuffer = ByteBuffer.allocate(headerSize - INITIAL_HEADER_SIZE);
+                headerBuffer.order(ByteOrder.LITTLE_ENDIAN);
+
+                read = stream.read(headerBuffer.array());
+                if (read != headerSize - INITIAL_HEADER_SIZE) {
+                    close();
+                    throw new RuntimeIOException(String.format("BCL %s has invalid header structure.", bclFile.getAbsoluteFile()));
+                }
+
+                byte bitsPerBasecall = headerBuffer.get();
+                byte bitsPerQualityScore = headerBuffer.get();
+
+                if (bitsPerBasecall != 2 && bitsPerBasecall != bitsPerQualityScore) {
+                    close();
+                    throw new RuntimeIOException("CBCL data not encoded in nibbles. (not currently supported)");
+                }
+
+                int numberOfBins = headerBuffer.getInt();
+
+                byte[] qualityBins = new byte[numberOfBins];
+                //each bin has a pair of 4 byte mappings
+                for (int j = 0; j < numberOfBins; j++) {
+                    int from = headerBuffer.getInt();
+                    int to = headerBuffer.getInt();
+                    qualityBins[j] = (byte) to;
+                }
+
+                int numTiles = headerBuffer.getInt();
+                TileData[] tileInfo = new TileData[numTiles];
+                for (int j = 0; j < numTiles; j++) {
+                    int tileNum = headerBuffer.getInt();
+                    int numClustersInTile = headerBuffer.getInt();
+                    int uncompressedBlockSize = headerBuffer.getInt();
+                    int compressedBlockSize = headerBuffer.getInt();
+                    tileInfo[j] = new TileData(tileNum, numClustersInTile, uncompressedBlockSize, compressedBlockSize);
+                }
+
+                boolean pfExcluded = headerBuffer.get() == 1;
+                cycleData[i] = new CycleData(version, headerSize, bitsPerBasecall, bitsPerQualityScore, numberOfBins, qualityBins, numTiles, tileInfo, pfExcluded);
+                this.streams[i] = stream;
+                this.streamFiles[i] = bclFile;
+                byteBuffer.clear();
+                headerBuffer.clear();
+            }
+
+
+        } catch (final IOException ioe) {
+            throw new RuntimeIOException(ioe);
+        }
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (queue == null || queue.size() == 0) {
```

Comment:
queue doesn't need to be a list any more since we only read a single byte that has already been translated from the nibble.